dlrover/python/diagnosis/inferencechain/inferenceoperator/check_training_hang_operator.py

```diff
@@ -77,17 +88,100 @@ def infer(self, inferences: List[Inference]) -> List[Inference]:
         ]
 
     def is_hang(self, diagnosis_data: List[DiagnosisData]):
-        hang_metric = []
+        logger.info(
+            "Hang detection start using diagnosis data, "
+            f"data number: {len(diagnosis_data)}, "
+            f"data size: {sys.getsizeof(diagnosis_data)}."
+        )
+        worker_hang_metric: Dict[int, List[Tuple[int, bool]]] = {}
         if not diagnosis_data:
             return False
 
         for data in diagnosis_data:
+            # filter hang metric
             each_metric = [
                 line
                 for line in data.data_content.splitlines()
                 if line.startswith(HANG_METRIC_PREFIX)
             ]
-            hang_metric.append(each_metric)
 
-        # TODO: implement the judgement
+            # if all local rank is hanged, tag worker hang
+            rank_hang_size = 0
+            is_worker_hang = False
+            for each_rank_metric in each_metric:
+                match = re.search(r"(\d+)(?!.*\d)", each_rank_metric)
+                if match and match.group(0) == "1":
+                    rank_hang_size += 1
+            if rank_hang_size == len(each_metric):
+                is_worker_hang = True
+
+            if data.node_rank not in worker_hang_metric:
+                worker_hang_metric[data.node_rank] = []
+            worker_hang_metric[data.node_rank].append(
+                (data.timestamp, is_worker_hang)
+            )
+
+        # hang detection rules:
+        # 1. 100% worker got hang metric
+        # 2. last for 10+ minutes
```

Comment:
Can we shorten the time 10 minutes. Torch 2.4 has set the default  timeout of nccl process group to 10min.

Reply:
5mins